### ISSUE TITLE
Should Dialyzer emit a warning in this code?

### ISSUE THREAD

**Issue by elbrujohalcon** [erlang/otp#6781]
I have a very simple module:

```erlang
-module my_mod.
-export [my_fun/1].

-spec my_fun(integer()) -> integer().
my_fun(X) -> X * 2.
```

If I run Dialyzer on it with `missing_return` warnings enabled, I get this warning:

```erlang
my_app/src/my_mod.erl:4:2: The success typing for my_mod:my_fun/1 implies that the function might also return
          float() but the specification return is
          integer()
```

I understand that the _success typing_ for the function includes floating point numbers, but the spec clearly states that it's only designed to receive integers. Is `missing_return` intentionally ignoring specs? Is there a way to configure Dialyzer so that it takes them into account?

**Comment by jhogberg** [comment]
No, the spec states that the function _will not return normally_ if given anything other than an integer. That does not mean that the function is not intended to accept floats, tuples, atoms, or anything else.

We have to assume that `X` can be anything coming in, because if `X` was constrained by the spec before "entering" the function we'd get super-misleading results in cases like:

```erlang
-spec contrived(integer()) -> {error, Reason};
contrived(X) when is_integer(X) ->
     {error, not_a_frobnitz};
contrived(X) ->
     server_function_that_never_returns(X).
```

If `X` is constrained to being an integer on entry, `server_function_that_never_returns/1` will not be reachable from here. Instead, `X` is constrained _on calls_ and analysis won't go past calls where `X` is not an integer.

> Should Dialyzer take them into account to avoid emitting such a warning?

I suppose we could try to reconcile them in some way, but it would be pretty involved and I'm not sure it would be worth the effort. It's not something I think we'll do in the near future at any rate :-/

**Comment by TD5** [comment]
> I understand that the _success typing_ for the function includes floating point numbers, but the spec clearly states that it's only designed to receive integers. Is `missing_return` intentionally ignoring specs? Should Dialyzer take them into account to avoid emitting such a warning?

My take on this is that `missing_return` and `extra_return` are handy options to alert you to _possible_ issues, but they don't fit into the idealised Dialyzer model of "if Dialyzer complains, there's definitely a path to a runtime issue". In practice, this means that I might toggle them on if I want to get some starting points when searching for potential issues, but generally keep them switched off, since they can generate false-positives.
